worker/controlsocket/worker.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package controlsocket
+
+import (
+	"net"
+	"net/http"
+	"runtime"
+
+	"github.com/juju/errors"
+	"github.com/juju/worker/v3"
+	"github.com/juju/worker/v3/catacomb"
+
+	"github.com/juju/juju/juju/sockets"
+)
+
+// logger is here to stop the desire of creating a package level logger.
+// Don't do this, instead use the one passed as manifold config.
+type logger interface{}
+
+var _ logger = struct{}{}
+
+// Logger represents the methods used by the worker to log information.
+type Logger interface {
+	Debugf(string, ...interface{})
+	Warningf(string, ...interface{})
+	Infof(string, ...interface{})
+}
+
+// Config represents configuration for the controlsocket worker.
+type Config struct {
+	state      State
+	logger     Logger
+	socketName string
+}
+
+// Validate returns an error if config cannot drive the Worker.
+func (config Config) Validate() error {
+	if config.state == nil {
+		return errors.NotValidf("nil state")
+	}
+	if config.logger == nil {
+		return errors.NotValidf("nil logger")
+	}
+	if config.socketName == "nil" {
+		return errors.NotValidf("empty socketName")
+	}
+	return nil
+}
+
+// Worker is a controlsocket worker.
+type Worker struct {
+	Config
+	catacomb catacomb.Catacomb
+	listener net.Listener
+}
+
+// NewWorker returns a controlsocket worker with the given config.
+func NewWorker(config Config) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	if runtime.GOOS != "linux" {
+		return nil, errors.NotSupportedf("os %q", runtime.GOOS)
+	}
+
+	l, err := sockets.Listen(sockets.Socket{
+		Address: config.socketName,
+		Network: "unix",
+	})
+	if err != nil {
+		return nil, errors.Annotate(err, "unable to listen on unix socket")
+	}
+	config.logger.Debugf("controlsocket worker listening on socket %q", config.socketName)
+
+	w := &Worker{
+		Config:   config,
+		listener: l,
+	}
+	err = catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.serve,
+	})
+	return w, errors.Trace(err)
+}
+
+func (w *Worker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+func (w *Worker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+// serve listens on the control socket and handles requests.
+func (w *Worker) serve() error {
```

Comment:
Rename to loop or run, as that's the going convention.

Reply:
`run` feels more appropriate here, there's no looping really